### PR TITLE
OP-1199 - Voucher Scheme new features (Subservices and subitems)

### PR DESCRIPTION
**Pull request by kevel-dev**

### CODE REVIEW

**Review by delcroip — APPROVED** [review]

**Comment by delcroip** [comment]
thanks you for the work,

@kevel-dev , @mngoe  please add test case for your features so I will know if someone break them

br

**Comment by kevel-dev** [comment]
@delcroip i have added some test cases

File: claim/utils.py [hunk 1]
```diff
@@ -62,8 +87,68 @@ def item_create_hook(claim_id, item):
...
+            )
+
+def service_update_hook(claim_id, service):
+    serviceLinked = service["serviceLinked"]
+    serviceserviceSet = service["serviceserviceSet"]
+    service.pop('serviceLinked', None)
+    service.pop('serviceserviceSet', None)
+    ClaimServiceId = ClaimService.objects.filter(claim=claim_id, service=service["service_id"]).first()
+    if(serviceLinked):
+        for serviceL in serviceLinked:
```
Comment (delcroip) [inline claim/utils.py:128]:
```suggestion
    serviceLinked = service["serviceLinked"]
    serviceserviceSet = service["serviceserviceSet"]
    service.pop('serviceItemSet', None)
    service.pop('serviceserviceSet', None)
    ClaimServiceId = ClaimService.objects.filter(claim=claim_id, service=service["service_id"]).first()
    if(serviceItemSet):
        for serviceL in serviceItemSet:
```

File: claim/validations.py [hunk 2]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
+                if ClaimConfig.native_code_for_services == False:
+                    try:
+                        if claim_detail.service.packagetype == 'F':
+                            service_price = claim_detail.service.price
+                            if claim_detail.price_adjusted is not None:
+                                print("compare ", claim_detail.price_adjusted, " and ", service_price)
+                                if claim_detail.price_adjusted > service_price:
+                                    set_price_adjusted = service_price
+                            else:
+                                print("compare ", claim_detail.price_asked, " and ", service_price)
```
Comment (delcroip) [inline claim/validations.py:1145]:
```suggestion
                                logger.debug("compare ", claim_detail.price_asked, " and ", service_price)
```

File: claim/validations.py [hunk 3]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
             if claim_detail.price_approved is not None:
                 set_price_adjusted = claim_detail.price_approved
+            if claim_detail.price_origin == ProductItemOrService.ORIGIN_CLAIM:
+                set_price_adjusted = claim_detail.price_asked
+                if ClaimConfig.native_code_for_services == False:
+                    try:
+                        if claim_detail.service.packagetype == 'F':
+                            service_price = claim_detail.service.price
+                            if claim_detail.price_adjusted is not None:
+                                print("compare ", claim_detail.price_adjusted, " and ", service_price)
```
Comment (delcroip) [inline claim/validations.py:1141]:
```suggestion
                                logger.debug("compare ", claim_detail.price_adjusted, " and ", service_price)
```

File: setup.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 setup(
     name='openimis-be-claim',
-    version='1.3.0',
+    version='1.3.2',
```
Comment (delcroip) [inline setup.py:12]:
```suggestion
    version='1.4.0',
```

File: claim/gql_mutations.py [hunk 5]
```diff
@@ -809,12 +833,55 @@ def async_mutate(cls, user, **data):
...
                     all_rejected = False
             services = data.pop('services') if 'services' in data else []
+            claimed = 0
+            ClaimServiceElts = []
             for service in services:
                 service_id = service.pop('id')
+                service_linked = service.pop('serviceLinked', [])
+                print("service_linked ", service_linked)
+                serviceserviceSet = service.pop('serviceserviceSet', [])
+                print("serviceserviceSet ", serviceserviceSet)
```
Comment (delcroip) [inline claim/gql_mutations.py:843]:
```suggestion
                logger.debug("serviceserviceSet ", serviceserviceSet)
```

File: claim/gql_mutations.py [hunk 6]
```diff
@@ -809,12 +833,55 @@ def async_mutate(cls, user, **data):
...
                 claim.services.filter(id=service_id).update(**service)
+                if ClaimConfig.native_code_for_services == False:
+                    for claim_service_service in serviceserviceSet:
+                        claim_service_code = claim_service_service.pop('subServiceCode')
+                        claim_service = claim.services.filter(id=service_id).first()
+                        if claim_service:
+                            service_elt = Service.objects.filter(code=claim_service_code).first()
+                            if service_elt:
+                                claim_service_to_update = claim_service.claimlinkedService.filter(service=service_elt.id)
+                                print("claim_service_to_update ", claim_service_to_update)
```
Comment (delcroip) [inline claim/gql_mutations.py:853]:
```suggestion
                                logger.debug("claim_service_to_update ", claim_service_to_update)
```

File: claim/gql_mutations.py [hunk 7]
```diff
@@ -809,12 +833,55 @@ def async_mutate(cls, user, **data):
...
+                                    claim_service_to_update.update(**claim_service_service)
+                            ClaimServiceElts.append(claim_service)
+                    for claim_service_item in service_linked:
+                        claim_item_code = claim_service_item.pop('subItemCode')
+                        claim_service = claim.services.filter(id=service_id).first()
+                        if claim_service:
+                            item_elt = Item.objects.filter(code=claim_item_code).first()
+                            if item_elt:
+                                claim_item_to_update = claim_service.claimlinkedItem.filter(item=item_elt.id)
+                                print("claim_item_to_update ", claim_item_to_update)
```
Comment (delcroip) [inline claim/gql_mutations.py:869]:
```suggestion
                                logger.debug("claim_item_to_update ", claim_item_to_update)
```

File: claim/validations.py [hunk 8]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
+                                if claim_detail.price_adjusted > service_price:
+                                    set_price_adjusted = service_price
+                            else:
+                                print("compare ", claim_detail.price_asked, " and ", service_price)
+                                if claim_detail.price_asked > service_price:
+                                    set_price_adjusted = service_price
+                    except:
+                        print("This it an item element")
```
Comment (delcroip) [inline claim/validations.py:1149]:
```suggestion
                        logger.warning("This it an item element")
```

File: claim/validations.py [hunk 9]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
+                    try:
+                        contunue_service_check = True
+                        if claim_detail.service.packagetype == 'P':
+                            service_services = ServiceService.objects.filter(servicelinkedService=claim_detail.service.id).all()
+                            claim_service_services = ClaimServiceService.objects.filter(claimlinkedService=claim_detail.id).all()
+                            if len(service_services) == len(claim_service_services):
+                                for servservice in service_services:
+                                    for claimserviceservice in claim_service_services:
+                                        if servservice.service.id == claimserviceservice.service.id:
+                                            print("comparing serviceservice qty ", servservice.qty_provided,
```
Comment (delcroip) [inline claim/validations.py:1162]:
```suggestion
                                            logger.debug("comparing serviceservice qty ", servservice.qty_provided,
```

File: claim/validations.py [hunk 10]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
+                                                set_price_adjusted = 0
+                                                contunue_service_check = False
+                                                break
+                                    if contunue_service_check == False:
+                                        break
+                            else:
+                                # user misconfiguration !
+                                set_price_adjusted = 0
+                                contunue_service_check = False
+                            print("set_price_adjusted after service check ", set_price_adjusted)
```
Comment (delcroip) [inline claim/validations.py:1174]:
```suggestion
                            logger.debug("set_price_adjusted after service check ", set_price_adjusted)
```

File: claim/validations.py [hunk 11]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
+                                contunue_item_check = True
+                                service_items = ServiceItem.objects.filter(servicelinkedItem=claim_detail.service.id).all()
+                                claim_service_items = ClaimServiceItem.objects.filter(claimlinkedItem=claim_detail.id).all()
+                                print("service_items: ", service_items)
+                                print("claim_service_items: ", claim_service_items)
+                                if len(service_items) == len(claim_service_items):
+                                    for serviceitem in service_items:
+                                        for claimservicesitem in claim_service_items:
+                                            if serviceitem.item.id == claimservicesitem.item.id:
+                                                print("comparing serviceitem qty ", serviceitem.qty_provided,
```
Comment (delcroip) [inline claim/validations.py:1185]:
```suggestion
                                logger.debug("service_items: ", service_items)
                                logger.debug("claim_service_items: ", claim_service_items)
                                if len(service_items) == len(claim_service_items):
                                    for serviceitem in service_items:
                                        for claimservicesitem in claim_service_items:
                                            if serviceitem.item.id == claimservicesitem.item.id:
                                                logger.debug("comparing serviceitem qty ", serviceitem.qty_provided,
```

File: claim/validations.py [hunk 12]
```diff
@@ -1132,13 +1131,73 @@ def process_dedrem(claim, audit_user_id=-1, is_process=False):
...
+                                                    contunue_item_check = False
+                                                    break
+                                        if contunue_item_check == False:
+                                            break
+                                else:
+                                    # user misconfiguration !
+                                    set_price_adjusted = 0
+                                print("set_price_adjusted after items check ", set_price_adjusted)
+                    except:
+                        print("This is a ClaimItem element, not a ClaimService")
```
Comment (delcroip) [inline claim/validations.py:1198]:
```suggestion
                                logger.debug("set_price_adjusted after items check ", set_price_adjusted)
                    except:
                        logger.debut("This is a ClaimItem element, not a ClaimService")
```

**Review by delcroip — COMMENTED** [review]
Overall:
- do not use print but logger instead (now logger show in console by default, checkout the latest assembly)
- why ServiceLinked and not ServiceItemSet ?

on the actual function I have no issue, I can merge it but I'd like to have a feefback on my second point

File: claim/gql_mutations.py [hunk 13]
```diff
@@ -809,12 +833,55 @@ def async_mutate(cls, user, **data):
...
+                                print("claim_service_to_update ", claim_service_to_update)
+                                if claim_service_to_update:
+                                    qty_asked = claim_service_service.pop('qty_asked', 0)
+                                    price_asked = claim_service_service.pop('price_asked', 0)
+                                    claim_service_service['qty_displayed'] = qty_asked
+                                    price = qty_asked * price_asked
+                                    claimed += price
+                                    claim_service_to_update.update(**claim_service_service)
+                            ClaimServiceElts.append(claim_service)
+                    for claim_service_item in service_linked:
```
[the remark this answers is not in the document]
Comment (kevel-dev) [reply claim/gql_mutations.py:862]: @delcroip  i think here you did not changed the variable service_linked to service_Item

**Comment by kevel-dev** [comment]
Thakns for your advices. I think all your changes are good (except the one i commented regarding the variable service_linked)

**Pull request closed without merging**